ocaml/xapi/helpers.ml

```diff
@@ -1234,3 +1234,122 @@ let host_supports_hvm ~__context host =
+  val restart : __context:Context.t -> accept:string -> unit
+end = struct
+
+  let cert = !Xapi_globs.server_cert_path
+
+  (** protect ourselves from concurrent writes to files *)
+  module Config : sig
+    (** create or update stunnel config file *)
+    val update : accept:string -> unit
+    (** calls `generate_ssl_cert` bash script
+      * TODO: don't generate certificates in bash *)
+    val generate_pem : __context:Context.t -> force:bool -> unit
+  end = struct
+    let xapi_stunnel_conf = "/etc/stunnel/xapi.conf"
+    let m = Mutex.create ()
+    let current_accept = ref None
+
+    let rewrite_xapi_ssl_config_file ~accept =
+      let fips = match Sys.getenv_opt "CC_PREPARATIONS" with
+                 | Some "true"   -> true
+                 | Some _ | None -> false
+      in
+      let conf_contents = String.concat "\n" [
+          "; autogenerated by helpers.ml"
+        ; Printf.sprintf "fips = %s" (if fips then "yes" else "no")
+        ; "pid = /var/run/xapissl.pid"
+        ; "socket = r:TCP_NODELAY=1"
+        ; "socket = a:TCP_NODELAY=1"
+        ; "socket = l:TCP_NODELAY=1"
+        ; "socket = r:SO_KEEPALIVE=1"
+        ; "socket = a:SO_KEEPALIVE=1"
+        ; (match Sys.getenv_opt "STUNNEL_IDLE_TIMEOUT" with
+          | None    -> "; no idle timeout"
+          | Some x -> Printf.sprintf "TIMEOUTidle = %s" x)
+        ; "debug = authpriv.5"
+        ; ""
+        ; "[xapi]"
+        ; Printf.sprintf "accept = %s" accept
+        ; "connect = 80"
+        ; Printf.sprintf "cert = %s" cert
+        ; Printf.sprintf "ciphers = %s" Xcp_const.good_ciphersuites
+        ; "curve = secp384r1"
+        ; "TIMEOUTclose = 1"
+        ; "options = CIPHER_SERVER_PREFERENCE"
+        ; "sslVersion = TLSv1.2"
+        ]
+      in
+      let len = String.length conf_contents in
+      Stdext.Unixext.atomic_write_to_file xapi_stunnel_conf 0o0600
+        (fun fd -> let (_: int) = Unix.single_write_substring fd conf_contents 0 len in ())
+
+    let update ~accept =
+      Mutex.execute m (fun () ->
+        match !current_accept with
+        | Some current_accept when current_accept = accept -> ()
+        | None | Some _ -> begin
+          current_accept := Some accept;
+          rewrite_xapi_ssl_config_file ~accept
+        end
+      )
+
+    let generate_pem ~__context ~force =
+      Mutex.execute m (fun () ->
+        if force || not (Sys.file_exists cert) then begin
+          let hostname = Unix.gethostname () in
+          let cn =
+            if Astring.String.is_prefix ~affix:"localhost" hostname then
+              get_management_ip_addr ~__context
+            else if Astring.String.is_infix ~affix:"." hostname then
+              None
+            else
+              get_management_ip_addr ~__context
+          in
+          D.debug "Helpers.Stunnel.gen_pem_if_not_exist: writing pem file";
+          let (_: string) = call_script "/opt/xensource/libexec/generate_ssl_cert" [cert; Option.value ~default:hostname cn]
```

Comment:
This should also go into globs. See how `xapissl` is done as an example. Also, we should never hardcode paths like`/opt/xensource/libexec`